### ISSUE TITLE
GitHub Issues Tracker Change

### ISSUE THREAD

**Issue by turtletooth** [opentoonz/opentoonz#1207]
I had a great chat with @ideasman42 tonight about the issues tracker here on GitHub. Here are the conclusions we reached:

Many of the bugs and crash reports are getting lost in the sea of feature requests or random issues that the developers can't do anything about.  

There are few active developers currently and having pages and pages of feature requests or non-critical issues makes it harder to identify what needs to be worked on.  (I am guilty of making my share of feature requests, but my perspective is changing)

The issues tracker here needs to be a place where developers can clearly see what needs to be done (without having to filter through labels).

It would be nice if feature requests could be moved to another site to be voted on so user voices can be heard, but github should contain actionable issues that developers need to/are willing to work on.  There are a ton of great ideas for feature requests but if no one is willing to take it on, it is just going to sit here and clog up the tracker.

So here is the proposal:

- Limit issues here to actual bugs/crashes or items that a developer is working on.  (This is a similar policy that Blender employs)
- Point users posting other issues to the Google Groups page or another site if there is a better option. (There is already a feature request page on Google Groups)
- Clean up the Google Groups feature request page to keep it focused and realistic. ( @shun-iwasawa I don't mind doing some work there if there is a way to give me permissions)
- Emphasize that the issues tracker here is for issues that need developer attention or are being actively worked on by a developer.

I know this sounds harsh but I think it would help make the issue tracker more realistic for the few developers that are working on this.

**Comment by jpturcotte** [comment]
If I may, I would like to play devil's advocate here.
Seeing how some in the community are still having trouble negotiating the different platforms to report bugs, ask for features or ask for help with the software (check out the subreddit for multiple examples), this might be a drastic change that fails to address the underlying cause: lack of communication. Drafting clear guidelines and relentlessly disseminating them could bring out the same results without risking confusing the recalcitrant users even further and dropping engagement. Else, you might just end up adding a second list of items to sift through in a different location to a still-disorganized bug tracker. Correctly and systematically labelling all 300+ issues would likely only take a day or two full time, and that is a mostly clerical task that could be delegated by the devs. In fact, I am willing and have the free time to do it. Once done, it would just be a matter of enforcing and keeping on top. If that fails to help you, then you might consider splitting the issues.
All that being said, I will back whatever decision you make because I trust in your judgement. I just wanted to make sure the both of you had considered all the possible repercussions of such a move.

**Comment by adamearle** [comment]
I wonder if a "Google form" at the end of every month could be published for users in the way of feature request for people to vote on. Maybe a discord meeting once a month for the feature request form? This way users would know they have a deadline to put a proposal in by a certain date the form is updated and then finally made available for voting. 

This maybe good over a more extended period of time as well because you will be to chart what are the overarching themes in the community as well. People would be submitting more so **"Feature Request Proposals**" so they would have to put more effort into what they are asking for with a clear message.

A community note on what makes a good proposal might be needed.

A problematic opportunity might be that some request are very similar to others, so they should be merged. Plus there needs to be a way the user can quickly and clearly search so that user can step out of their own ego and merge their request with another users if they need to or they can work on feature request together. Discord OpenToonz group maybe good for people to start connecting this way.

The discord Opentoonz has certain # tags setup so that users can treat that space as somewhere they can go and talk about something very specific ie. the Xsheet 

We could also encourage the users and promote the idea that this is where they could come together to start chatting working on feature request ie Think tanks for feature request

I wonder if something like a feature request wiki might help where feature request can be extended upon or added too. The wikki could be broken up into clear ideas for feature request such as 

**Animation:**
1. Keyframing
2. Rigging
3. Plastic Tool
4. Edit Tool
5. Pegs
6. Workflow
7. Audio

**Timelines:**
1. XSheet
2. Levels (Frame library)
3. Function editor
4. Curve editor

**Drawing:**
1. Vector 
2. Raster
3. Convert to Vector

**Rendering:**
1. Camera
2. FX
3. Export

I can't help but feel though users are going to run into the problem where topic cross over and the request is half in one group and half in the other. I am managing the OpenToonz Discord at the moment  and www.YourOpenToonz.com plus the Google+ community (It's tiny)
https://plus.google.com/communities/117328766497043417332 but I am happy to accept a helping hand to perhaps push some of these ideas across.

Love to hear your thoughts this is a good conversation to be having.

**Comment by adamearle** [comment]
Let not forget the Facebooks, Redits and Tweets plus also other misc forums of importance. It could actually be a good push for collecting more Devs, Artist and Studios using OT.

**Comment by adamearle** [comment]
This is here if you want to use it https://docs.google.com/drawings/d/1Rr81BgpUCABqV6ZDXPXoDzVqfTz7AbToBYMm6_MpwVI/edit

**Comment by adamearle** [comment]
here is a demo form https://docs.google.com/forms/d/1ZVwuyYCDWpTGUEqUlk6tXeoGQMTT73aacvPp82EauRA/edit

**Comment by turtletooth** [comment]
@jpturcotte @adamearle Thanks for the thoughts.  There are some good ideas there.

@jpturcotte I agree that it is important for it to be clear where users should post things related to certain issues.  There is room for improved communication there.  

On this particular topic though, I think it is important to remove things from the issues here that may never see the light of day.  Over 1/3 of the issues here are feature requests.  Unless someone steps up and says that they will take something on, it will only sit there adding noise to the true problems/issues.  LMMS has  the same problem.

For now, I think all feature requests should be on the Google Group with it being made very clear that just requesting something is no guarantee that it will happen.  I only have time to work on bugs or features that will directly benefit my students, @shun-iwasawa is doing a great job trying to implement things that Japanese studios are asking for and fixing bugs, @blackwarthog is doing some great things also.  But time is finite and since this is open source with no paid developers (except for bounties and users hiring devs for specific features), the issues tracker here needs to reflect the work that has a strong chance of happening (bugs/features that devs need).  Organizing the tracker is here is useful but it still creates noise/distraction from what the urgent issues are.  @shun-iwasawa and I both check the Google Group regularly so it's not like conversations over there won't be heard.  But the issues tracker here has become a general forum for all things OpenToonz instead of focusing on what can happen in code.  

@adamearle I like the idea of a Google form or something like that.  Blender has rightclickselect.com where users can submit and vote on features.  That is open source, so if someone has the time to set up a server and mysql backend to run it, we could have something like that.  Again the reality is that unless a dev wants to build something, many if not most feature requests won't happen (at least for a long time).

**Comment by RodneyBaker** [comment]
I'm not a developer so take that into consideration with regard to the following thoughts on 'new features'...

Voting for features doesn't seem like a good idea... unless,,, the features under consideration are all at least potentially going to be developed by those who will be involved in coding that feature and can cast a vote because their time and effort is at stake.

It seems the bottom line is that (as has been previously stated) there must be a programmer willing to champion a feature.  

I'm hesitant to suggest dropping feature requests from github because they've worked so well in the past.  The free flow of ideas is what continues to improve OpenToonz as developers are inspired and consider how those ideas can be transformed into code.  Several of my suggestions have been implemented and I those features almost every time I open the program.  Having said that however, I do understand the organizational overhead (and clutter) that tons of feature requests generate however.  

I think I am agreement that feature requests *that are not actively being worked on by a programmer or beginning to do so* should be relegated to the Google group.  The Google group can use the additional traffic anyway and discussion of features can be as in depth as anyone cares to go with them.  :)

New features could be added to github but specifically only those being actively worked on.  An example of this would be those that pop into existence suddenly by @shun-iwasawa.  He has the ear of the Japanese studios.  

The difficult part of this is the policing of the github entries.  Someone... or several someones... would have to get a little heavy handed and (as gently as possible) point those that post feature requests not actively being worked on to the Google group features forum where the feature can be actively developed at the user level in hopes of a champion emerging that is willing and able to take on the challenge of writing the code.  In this way it up to the feature requester to pitch their idea and put it in the best light possible in hope that it might graduate to being an acknowledged feature request.

 Disclaimer:  I'm an admin on another forum where something similar occurs but there is really only one programmer... so it is quite different than an open source project.  There are several places where users can discuss ideas for features but there is really only one where 'official feature requests' get considered and the programmer is the one to assign that to himself.  The system used there is [Mantis](http://www.mantisbt.org/).

What we could do (and by 'we' I suppose I mean mostly me) is point any feature request that appears here without programmer attached to a topic in the OpenToonz Google group where the feature can be discussed.  The big question would be what to classify the entry as here in github.  Can it be tagged as 'acknowledged' rather than 'closed'?  The issue with tagging as 'closed' is that designation can seem to imply implementation.  

Ideas are the life blood of development so we (collectively) need to continue to encourage feature requests.  This is important also because users want to be able to share their experience and workflow (real and idealized).  How to best approach that so that developers have a clearer path toward their own personal goals... now that is the question!

Respectively submitted... and in awe of those who code.  :)

**Comment by turtletooth** [comment]
@RodneyBaker Thank you for your thoughts.  I agree that feature requests that have a programmer and are not orphaned should stay here.  

Regarding closed- unfortunately that is a GitHub implementation.  There is no way of removing an issue from the list without closing it.  I think a comment related to the reason for closing would be sufficient to clarify it.

**Comment by RodneyBaker** [comment]
I think I'd be willing to take on trying to promote *user led development* discussions in the Google Groups arena.  There is a lot of gray area there because I'm not an admin but when last I checked even common members had the ability to reclassify topics (that is to say... tag topics so they fall under an appropriate topic).  I spent considerable time reclassifying topics shortly after the Google Groups got off the ground and nothing was organized.  That was... um... fun.  (Seriously though, I learned a ton by just reading all of those posts!) :)

But... this illustrates that everyone can be involved in 'development' at some level.  We all want OpenToonz to be the absolute best program possible and hopefully (for users) that is because we want to actually use it (or in the case of instructors/teachers) assist others so they can use it.

At any rate, I'm more than willing to volunteer my time to open a Google Group topic for any feature request that is closed here on github.  The idea being that if closed here the feature is being opened for further user led development and discussion.  Then programmers that are looking to take on a challenge can simply look to the Googles Groups discussions to find a feature they think worthy of championing.

I do think a standard closing remark would be appropriate for those feature requests that are closed here.  It might be something as simple as "Closed.  See:  " https://groups.google.com/forum/#!categories/opentoonz_en/feature-requests"

**Comment by turtletooth** [comment]
@RodneyBaker That is a very reasonable proposal.  You have already been a great voice on the Google Group.  The Google Group has been a little quiet recently and it would be good to build up the community there.  

If you are interested a good place to start might be taking the feature requests that have the most comments here and starting a page for them over there.  

I don't know if @shun-iwasawa can assign privileges over there to allow for greater sorting but that would be useful.  It might even be nice to have a pinned page at the top of the feature request section to give a one line summary of each requested feature.

**Comment by ideasman42** [comment]
re: @jpturcotte 

> If I may, I would like to play devil's advocate here.
> Seeing how some in the community are still having trouble negotiating the different platforms to report bugs, ask for features or ask for help with the software (check out the subreddit for multiple examples), this might be a drastic change that fails to address the underlying cause: lack of communication.

Why assume there *should* be communication to begin with?

If the user-developer ratio is 1:100 or 1:1000... *(just a guess, but it may be close)*, then its not reasonable to expect developers to be keeping up responding to requests from the user community.

> Drafting clear guidelines and relentlessly disseminating them could bring out the same results without risking confusing the recalcitrant users even further and dropping engagement. Else, you might just end up adding a second list of items to sift through in a different location to a still-disorganized bug tracker.

At least in this case its not interfering with the process of users reporting errors and having them fixed.

> Correctly and systematically labelling all 300+ issues would likely only take a day or two full time, and that is a mostly clerical task that could be delegated by the devs. In fact, I am willing and have the free time to do it. Once done, it would just be a matter of enforcing and keeping on top. If that fails to help you, then you might consider splitting the issues.

The issue is this isn't a one-off task, its continued maintenance to support this.

**Comment by RNavega** [comment]
From the point of view of a random user, GitHub looks like the best place to post feature requests since it's intimately linked to the official product page ( opentoonz.github.io/e/ ). Most people will try to give their suggestion right here since they'll think it's more likely to reach a developer and get implemented.

But for it to work in a way that doesn't "pollute" the main repository, a secondary repository can be created -- something like "opentoonz_ideas", a repository whose sole purpose is to collect ideas\feature requests in the issue tracker for junior or senior developers to pick and choose when they feel like it. That place can safely become a swamp.
This would require the addition of some text to the [Community section of the readme file](https://github.com/opentoonz/opentoonz#community) that explains the following:

> - For FEATURE REQUESTS and IDEAS, please use the [opentoonz_ideas issue tracker](https://github.com/opentoonz/opentoonz_ideas/issues).
> - For CRASHES and BUG REPORTS, please use the [opentoonz issue tracker](https://github.com/opentoonz/opentoonz/issues).
>
> Thank you for understanding!

This doesn't stop people from making the occasional mistake, but it should redirect much of the flow.

**Comment by turtletooth** [comment]
@RNavega I think opening a second repo on GitHub might me more confusing than just using the Google Group.  GitHub by design is for programmers.  Asking users to hang out on a site for programmers doesn't make as much sense to me as creating a community for users in the Google Group.  @shun-iwasawa and I already check the Google Group regularly and I don't want to have another area to watch.

**Comment by artisteacher** [comment]
I can definitely see how the current system is unsustainable with the number of active developers. I think keeping the bugs separate from feature requests that are not being developed makes sense. I do have a couple questions though:

If feature requests are limited to the google group, would there still be a process for bounties? Thanks to bountysource & etc. we have the mypaint brushes coming.

Also, what about things that are not necessarily bugs, but are areas of inconsistency in the program? Should they be reported as issues or feature requests?

**Comment by turtletooth** [comment]
In the case of bounties, I would personally be fine for them to be issues here as they are much more likely to get developed.  It would also need to be taken on a case by case basis over time, though- for example if a bounty has been open for six months or a year and is no where close to getting its funding, it may be pointless to leave it open.

I think that inconsistencies or areas where the program doesn't behave as would be expected make sense to post here.  They are often not too difficult to code and should be brought to the attention of developers directly.

The biggest things I would like to move to the Google Group are the things that are wish list items with no active developer.

**Comment by ideasman42** [comment]
I don't have a strong opinion either way - but making an exception for bounties could be abused.

Eg, someone makes many very large feature requests and puts a relatively small bounty on each.

Whatever the case - for as long as its not being abused, having an exception for bounties could work, closing them after a time period when they're not met.

**Comment by shun-iwasawa** [comment]
@RodneyBaker @turtletooth This is a quick note that I just assigned both of you as the "Community Manager" role of the Google Group. It is almost identical to the Manager role, except that the Community Manager has no access for assigning another users to any role. Please let me know if there is any problem.

**Comment by turtletooth** [comment]
Thanks @shun-iwasawa

**Comment by morevnaproject** [comment]
My two cents about the situation.

Moving feature requests from GitHub to Google Groups closes the possibility to put bounties for implementing them. Because Bountysource doesn't supports Google Groups.

Google Groups have disadvantage, because they are not allow "@"-mentions, like GitHub. With GitHub users usually put mention to ask me or someone else if it is possible to implement something through bounty. With Google Groups there are no such possibility,

> Many of the bugs and crash reports are getting lost in the sea of feature requests or random issues that the developers can't do anything about.

The re is a feature of GitHub, that is especially targeted on organizing issues. It is called Project Boards - https://help.github.com/articles/about-project-boards/

I am highly recommend to learn how to use it. It is simple, but really powerful feature to organize and coordinate development. And it is more powerful than tags. You can see how it is used for Synfig (https://github.com/synfig/synfig/projects/1) and MyPaint (https://github.com/mypaint/mypaint/projects and https://github.com/mypaint/libmypaint/projects/1).

**Comment by morevnaproject** [comment]
> Voting for features doesn't seem like a good idea...

I am not agree. For my team the count of user votes is a sign that we consider when discussing next feature to implement (currently we are estimating the count of people participating in discussion as count of interested people). This is also correlates with the chances to get funding for particular feature.

There is nice voting feature provided by UserEcho service - they limit amount of votes for each user. That makes users spread their voices carefully.

**Comment by turtletooth** [comment]
@morevnaproject I understand your points but keeping around 100 feature requests open without any guarantee that they will ever be worked on is somewhat pointless.  If a user is willing to put a bounty on a feature or a developer is willing to take it on, it certainly makes sense to have here, but until then, it is just noise in an already large collection of needed fixes.  

I do think looking into project boards makes sense.

**Comment by RodneyBaker** [comment]
@morevnaproject 
Project boards... definitely worth looking into.

Funding features (via bounty) shouldn't be a problem.  If anyone opts to initiate a bounty for a feature it can be assumed development is being initiated (i.e. the feature has an active champion) and they can start an entry here on OT github.  The idea being that it is actively being developed/championed to the point where someone has taken the time to initiate the funding cycle.  Of course, this doesn't mean it'll get funded so... whomever is championing that bounty needs to keep tabs on it's progress and where there is no interest acknowledge such by closing the github entry.  It's important to keep a good name so leaving orphaned issues/posts/PRs etc. easily conveys the interest of parties involved.  It's not hard to see what issues are truly important to people.

As for voting... you've case is a bit different than most.  You guys/gals are a 'team' and as such must decide where your priorities will be.  It makes perfect sense that your team would vote on various matters.  Should this icon be gray or blue... let's take an informal vote on it! 

Aside:  The true value of voting is getting the decision made rather than leaving it open to endless speculation and debate.  If the wrong candidate gets voted in that'll be apparent soon enough and appropriate changes can be made. 

Just in case someone might think I'm opposed to 'voting'... I"m a big fan of surveys and such especially where they can aid in communication.  Some care does have to be made to account for biases, foregone conclusions and such.  Of course this is where Team Leads get to be leaders.  They are in the best possible position to know the members of their team.  Guys... gray or blue icon... what do you think?  

I'd go on into additional detail about why voting for features isn't generally a good idea (in the larger scheme of things) but I don't think we really need to go there.  Good ideas will always return.

Remind me to share my thoughts on ideal voting/approval systems some day.


Bottom line:  I don't think other devs will have any issue with you opening PRs on the OpenToonz github.  You are actively developing OT and that activity speaks for itself.

**Comment by RodneyBaker** [comment]
Thanks @shun-iwasawa   :)

**Comment by adamearle** [comment]
Just throwing this out there but maybe something like Trello might be able to help
Personally: 
I love bounties
Blender right click just seems like a space to keep people quiet - is it even categorized and can run searches across the site
communication is key IMO and setting it up so people can connect easily is important.

That's all I got for now

sorry guys busy week

Adam

**Comment by adamearle** [comment]
OpenToonz Trello https://trello.com/b/Rd0THVfS/opentoonz

**Comment by adamearle** [comment]
Alternatives https://opensource.com/alternatives/trello

**Comment by RNavega** [comment]
I suggest changing the ['feature request' paragraph of the guidelines for OT](https://github.com/opentoonz/opentoonz/blob/master/CONTRIBUTING.md#features) to reflect this new change in how feature requests should be posted on Google Groups etc.

**Comment by blurymind** [comment]
The only problem with this change is that actual active developers will never see the feature requests  - because they dont actually go to the google plus boards

**Comment by blurymind** [comment]
Problem two with the change is integration - when a feature request is mentioned in a pull request - github shows that at both places. That is very useful and will be lost if you rely on external websites for feature requests

**Comment by ideasman42** [comment]
> The only problem with this change is that actual active developers will never see the feature requests - because they dont actually go to the google plus boards

Why do you assume this? If users make good well thought thorough proposals, and developers are available, I'm sure developers will find out about them.

> Problem two with the change is integration - when a feature request is mentioned in a pull request - github shows that at both places. That is very useful and will be lost if you rely on external websites for feature requests

Before tasks are worked on proposals can be moved to an issue in github to finalize - if its needed.

**Comment by blurymind** [comment]
I assume it based on experience with other open source projects, where the shift caused both less feature requests and participation to implement them.

But it has worked well for projects such as blender and it could for opentoonz. I understand the need to clean up the tracker - so as to give priority to bug fixes

**Comment by mcjadski** [comment]
I in principle agree with this decision, at least for now; once the team becomes larger, and once the glitches are under control, the feature requests can have a bigger role.  

However, I noticed that #1094 was closed because of this new policy.  I think there has to be a distinction between "I've got this cool way for how to make the program more convenient" and "The program isn't working the way it should (even if it's technically not a glitch)."

**Comment by turtletooth** [comment]
@mcjadski There is already a PR in for #1094.  Please see #1250

**Comment by mcjadski** [comment]
@turtletooth Oh oops, didn't see that.  Thanks!!  =)

**Comment by Skrylar** [comment]
Except, this kind of thing is exactly why Github has a label system?
